### PR TITLE
audit: disallow github.com/org/repo/pull/n/commits/hash.patch?full_index=1

### PR DESCRIPTION
**Pull request by wdconinc**
GitHub patches of the format https://github.com/spack/spack/pull/1/commits/b4da28f71e2cef84c6e289afe89aa4bdf7936048.patch?full_index=1 do not have a full index and should be disallowed.

This PR detects these cases in audit.

### CODE REVIEW

**Comment by wdconinc** [comment]
@spackbot fix style

File: lib/spack/spack/audit.py [hunk 1]
```diff
@@ -436,24 +440,27 @@ def _check_patch_urls(pkgs, error_cls):
                 if not isinstance(patch, spack.patch.UrlPatch):
                     continue
 
-                if re.match(github_patch_url_re, patch.url):
+                if re.match(github_unstable_pull_commits_re, patch.url):
+                    errors.append(
+                        error_cls(
+                            f"patch URL in package {pkg_cls.name} must not be a pull request commit",
```
Comment (wdconinc) [inline lib/spack/spack/audit.py]:
```suggestion
                            f"patch URL in package {pkg_cls.name} "
                            + "must not be a pull request commit",
```

File: lib/spack/spack/audit.py [hunk 2]
```diff
@@ -436,24 +440,27 @@ def _check_patch_urls(pkgs, error_cls):
...
+                    )
+                elif re.match(github_patch_url_re, patch.url):
                     full_index_arg = "?full_index=1"
                     if not patch.url.endswith(full_index_arg):
                         errors.append(
                             error_cls(
-                                "patch URL in package {0} must end with {1}".format(
-                                    pkg_cls.name, full_index_arg
-                                ),
+                                f"patch URL in package {pkg_cls.name} must end with {full_index_arg}",
```
Comment (wdconinc) [inline lib/spack/spack/audit.py]:
```suggestion
                                f"patch URL in package {pkg_cls.name} "
                                + f"must end with {full_index_arg}",
```

**Comment by wdconinc** [comment]
@spackbot fix style

**Comment by adamjstewart** [comment]
Out of curiosity, does this also affect PR patches like `https://github.com/pytorch/pytorch/pull/55063.patch?full_index=1`? I use those heavily in the PyTorch recipe. If not, maybe we could change some of these URLs to those instead. I'm always hesitant to blindly copy-n-paste patch files and add them to Spack for legal reasons.

**Comment by alalazo** [comment]
We have this repository: https://github.com/spack/patches Wondering if we should start using that, if there is no way to recover the patch reliably from the web?

**Comment by wdconinc** [comment]
> Out of curiosity, does this also affect PR patches like `https://github.com/pytorch/pytorch/pull/55063.patch?full_index=1`

As far as I can tell, this only affects the patch urls for individual commits of pull requests. Patch urls for an entire PR like the one you sent are stable (at least as long as the PR has been merged or no one adds to it or rebased it).

This honestly looks like an oversight/bug by GitHub and they may even be willing to fix it and provide a full index.

**Comment by wdconinc** [comment]
> As far as I can tell, this only affects the patch urls for individual commits of pull requests. 

I'll add that these commits live on a branch (often on a fork), and as long as that repo/branch exists, you can get a full index that works by referring to the commit on the branch. So:
- `https://github.com/pytorch/pytorch/pull/55063/commits/10a6a6a7e0a30b306f8580a869d3a0660ce54686.patch?full_index=1` bad
- `https://github.com/maxim-belkin/pytorch/commit/10a6a6a7e0a30b306f8580a869d3a0660ce54686.patch?full_index=1` good (from the point of view of having a full index... but bad for other reasons, like requiring maxim-belkin to continue having this fork)

Maybe the same GitHub PM as in https://github.com/orgs/community/discussions/46034#discussioncomment-4846112 would be able to comment, but I'll leave that summoning to someone else who may know if there are other communication channels between GitHub and Spack.

**Comment by tgamblin** [comment]
A few thoughts:

1. I dug into this and the reason this is bad:
    * `https://github.com/pytorch/pytorch/pull/55063/commits/10a6a6a7e0a30b306f8580a869d3a0660ce54686.patch?full_index=1`
    Is because it redirects you to:
    * `https://github.com/pytorch/pytorch/commit/10a6a6a7e0a30b306f8580a869d3a0660ce54686.patch`

    So really what is happening here is that whatever is doing the redirect on Github's side is not preserving `?full_index=1`. If you add it back to the redirect target, you get a full index.

    I think you should be using the second form regardless, as I don't know whether GitHub guarantees that it'll preserve the association between a pull request and a commit.  I think it would be good if the audit could suggest using, e.g.:
     * `https://github.com/pytorch/pytorch/commit/10a6a6a7e0a30b306f8580a869d3a0660ce54686.patch?full_index=1`

    Or just fix it in place (though I don't think we currently have `spack audit --fix` ... maybe add this to `spack style --fix`?)

2. RE @alalazo 
    > We have this repository: https://github.com/spack/patches Wondering if we should start using that, if there is no way to recover the patch reliably from the web?
    
    We mirror all these patches to mirror.spack.io, so we do currently preserve them already, without a lot of manual effort. They're fetchable there by their `sha256`.

**Comment by wdconinc** [comment]
Thanks!

> I think it would be good if the audit could suggest using, e.g.: `https://github.com/pytorch/pytorch/commit/10a6a6a7e0a30b306f8580a869d3a0660ce54686.patch?full_index=1`

Ok, I change that in the audit (not the `--fix` for now, but a suggestion), propagate it to the affected patch urls, remove the uploads, update the sha256s. Then this is unstuck, and at least flags the issue with a fix. Since there is no patch to upload, it doesn't relate to 2 anymore.

**Comment by wdconinc** [comment]
New audit suggestions:
```console
root@codespaces-3df160:/workspaces/spack# spack audit packages
PKG-DIRECTIVES: 10 issues found
1. patch URL in package dla-future must not be a pull request commit; instead use https://github.com/eth-cscs/DLA-Future/commit/efc9c176a7a8c512b3f37d079dec8c25ac1b7389.patch?full_index=1
    https://github.com/eth-cscs/DLA-Future/pull/1063/commits/efc9c176a7a8c512b3f37d079dec8c25ac1b7389.patch?full_index=1
2. patch URL in package miopen-hip must not be a pull request commit; instead use https://github.com/ROCm/MIOpen/commit/f60aa1ff89f8fb596b4a6a4c70aa7d557803db87.patch?full_index=1
    https://github.com/ROCm/MIOpen/pull/2276/commits/f60aa1ff89f8fb596b4a6a4c70aa7d557803db87.patch?full_index=1
3. patch URL in package mlpack must not be a pull request commit; instead use https://github.com/mlpack/mlpack/commit/183396e51a6771d5d2b43f22b0d2a9a91785e533.patch?full_index=1
    https://github.com/mlpack/mlpack/pull/3502/commits/183396e51a6771d5d2b43f22b0d2a9a91785e533.patch?full_index=1
4. patch URL in package openscenegraph must not be a pull request commit; instead use https://github.com/openscenegraph/OpenSceneGraph/commit/759620a3b7b787c960a7e414ba26ab5497817d40.patch?full_index=1
    https://github.com/openscenegraph/OpenSceneGraph/pull/1281/commits/759620a3b7b787c960a7e414ba26ab5497817d40.patch?full_index=1
5. patch URL in package protobuf must not be a pull request commit; instead use https://github.com/protocolbuffers/protobuf/commit/3039f932aaf212bcf2f14a3f2fd00dbfb881e46b.patch?full_index=1
    https://github.com/protocolbuffers/protobuf/pull/11032/commits/3039f932aaf212bcf2f14a3f2fd00dbfb881e46b.patch?full_index=1
6. patch URL in package py-astor must not be a pull request commit; instead use https://github.com/berkerpeksag/astor/commit/30059dac4eb832e58ab2109db84508b294ba366d.patch?full_index=1
    https://github.com/berkerpeksag/astor/pull/163/commits/30059dac4eb832e58ab2109db84508b294ba366d.patch?full_index=1
7. patch URL in package umpire must not be a pull request commit; instead use https://github.com/LLNL/Umpire/commit/47ff0aa1f7a01a917c3b7ac618e8a9e44a10fd25.patch?full_index=1
    https://github.com/LLNL/Umpire/pull/805/commits/47ff0aa1f7a01a917c3b7ac618e8a9e44a10fd25.patch?full_index=1
8. patch URL in package umpire must not be a pull request commit; instead use https://github.com/LLNL/Umpire/commit/2292d1d6078f6d9523b7ad0886ffa053644569d5.patch?full_index=1
    https://github.com/LLNL/Umpire/pull/816/commits/2292d1d6078f6d9523b7ad0886ffa053644569d5.patch?full_index=1
9. patch URL in package wrf must not be a pull request commit; instead use https://github.com/wrf-model/WRF/commit/4a084e03575da65f254917ef5d8eb39074abd3fc.patch?full_index=1
    https://github.com/wrf-model/WRF/pull/1888/commits/4a084e03575da65f254917ef5d8eb39074abd3fc.patch?full_index=1
10. patch URL in package wrf must not be a pull request commit; instead use https://github.com/wrf-model/WRF/commit/6087d9192f7f91967147e50f5bc8b9e49310cf98.patch?full_index=1
    https://github.com/wrf-model/WRF/pull/1888/commits/6087d9192f7f91967147e50f5bc8b9e49310cf98.patch
```

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
> We mirror all these patches to mirror.spack.io, so we do currently preserve them already, without a lot of manual effort. They're fetchable there by their sha256.

Sorry, I meant: "It's preferable to still have an https patch, but if we have to store huge patches, maybe move them in the other repo." Anyhow, the solution now seems much better than adding 7k lines of patches to the main repo.

**Pull request merged**